### PR TITLE
feat: uses verdaccio instead mock manually the registry

### PR DESCRIPTION
**Pull request by juanpicado**
I would like to file a proposal. 

I've noticed the mock registry was quite complicated to maintain and quite old, thus since I am Verdaccio maintainer https://verdaccio.org/ I decided to proceed to replace such implementation with the verdaccio registry.

https://www.npmjs.com/package/@verdaccio/how-to-npm

I have a PoC ready to use for code review.
```
npm install -g @verdaccio/how-to-npm
```
For the user nothing changes, all the problems are not affected at all, Verdaccio runs under the hood, in the same way, the previous implementation, it will reduce the amount of code to maintain, full npm registry experience and a UI where users can browse for more visual feedback about what is on the registry while they proceed with the tasks.

~In this PoC I'm using ` "verdaccio": "4.0.0-alpha.7"` since us bug-free and more updated with the current npmjs endpoints, but the latest version can be used also perfectly.~ The latest is already `v4`. 

Furthermore, new problems might be added since Verdaccio is adding constantly new endpoints as search, starts and soon tokens.

### CODE REVIEW

**Comment by juanpicado** [comment]
Would this be interesting for anyone ? Otherwise I would close this PR :(

**Comment by emzinnia** [comment]
@juanpicado Don't close it! This is awesome! I'll take a look at it, but it looks good

@watilde perhaps you might wanna review this as well?

**Comment by juanpicado** [comment]
Nice 🎉 . I'll update the dependency, it is a bit old already.

**Comment by watilde** [comment]
Hello! Thank you for your big effort. To help me understand more about the purpose of this PR as refactoring for better maintenance, I have a small comment since I'm wondering if this PR is adding more lines instead of reducing.
> +4,645 −257 

I see an improvement at `lib/registry.js`, but do we need to add more mock files to maintain as a tradeoff?

**Comment by juanpicado** [comment]
- The 4k lines are mostly because I commited the `package-lock.json` file, my bad.
-  The metadata (mock data) was there already, the diff is I added them prettyfied. 

If I use single line for metadata and then remove lock file, you will get more red than green.  

Let me know whether I should use single line, but IMHO, I would leave them prettified because are more readable, it is handy to see what's is inside, specially in code reviews.

File: problems/10-publish-again/index.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ exports.verify = function (args, cb) {
 
   var __ = this.__
   var pkg = require(path.join(shop.cwd(), 'package.json'))
-  var data = require(path.join(shop.datadir, 'registry', pkg.name, 'body.json'))
+  var data = require(path.join(shop.datadir, 'registry/store', pkg.name, 'package.json'))
```
Comment (ccarruitero) [inline problems/10-publish-again/index.js]: I think we should keep same style for similar lines. I prefer:
https://github.com/workshopper/how-to-npm/blob/964688c5b525cea41304cb5cc4aff251b96d5330/problems/11-dist-tag/index.js#L22

**Comment by ccarruitero** [comment]
I think would be better for review to move prettifier changes to other PR.

Aside other comments the code seems good.

File: problems/10-publish-again/index.js (hunk 1 above)
Comment (juanpicado) [reply problems/10-publish-again/index.js]: Agree, make sense, but I guess you meant `data` rather  `body` . Unless you want I change it also to `body`.

**Comment by juanpicado** [comment]
done 👍

File: problems/12-dist-tag-removal/index.js [hunk 2]
```diff
@@ -19,7 +19,7 @@ exports.verify = function (args, cb) {
   var pkg = require(path.join(cwd, 'package.json'))
   var name = pkg.name
 
-  var body = require(path.join(shop.datadir, 'registry', name, 'body.json'))
+  var body = require(path.join(shop.datadir, 'registry/store', name, 'package.json'))
```
Comment (ccarruitero) [inline problems/12-dist-tag-removal/index.js]: keep this line same style others too

File: assets/registry/store/@linclark/pkg/package.json [hunk 3]
```diff
@@ -1 +1 @@
...
...m":"2f10c9ae04e61a7efa9fb43097352808321705a0","tarball":"http://registry.npmjs.org/@linclark/pkg/-/pkg-1.0.1.tgz"},"maintainers":[{"name":"linclark","email":"lin.w.clark@gmail.com"}],"directories":{}},"1.0.2":{"name":"@linclark/pkg","version":"1.0.2","description":"An empty package for the how-to-npm workshopper to use.","main":"index.js","author":{"name":"Lin Clark"},"license":"ISC","_id":"@linclark/pkg@1.0.2","scripts":{},"_shasum":"d031778d174e04874a2c8eaba5928898a1d50292","_from":".","_npmVersion":"2.9.1","_nodeVersion":"0.10.35","_npmUser":{"name":"linclark","email":"lin.w.clark@gmail.com"},"dist":{"shasum":"d031778d174e04874a2c8eaba5928898a1d50292","tarball":"http://registry.npmjs.org/@linclark/pkg/-/pkg-1.0.2.tgz"},"maintainers":[{"name":"linclark","email":"lin.w.clark@gmail.com"}],"directories":{}}},"readme":"ERROR: No README data found!","maintainers":[{"name":"linclark","email":"lin.w.clark@gmail.com"}],"time":{"modified":"2015-05-21T20:16:45.507Z","created":"2015-05-21T16:34:50.033Z","1.0.0":"2015-05-21T16:34:50.033Z","1.0.1":"2015-05-21T16:36:33.580Z","1.0.2":"2015-05-21T20:16:45.507Z"},"author":{"name":"Lin Clark"},"license":"ISC","readmeFilename":"","_attachments":{}}
```
Comment (ccarruitero) [inline assets/registry/store/@linclark/pkg/package.json]: Not sure why this file is modified?

**Comment by ccarruitero** [comment]
@juanpicado thanks for all the work you put on this PR.

Aside my comments, just ensure assets files aren't modified if not necessary and please squash your commits.

**Comment by juanpicado** [comment]
@ccarruitero done 👍

**Comment by ccarruitero** [comment]
@juanpicado sorry for delay. Thanks for all the work you did.

I think the assets files should be just renamed, but there are some that are modified.
Also, check [this comment](https://github.com/workshopper/how-to-npm/pull/143#pullrequestreview-318553793) for the style in the require line

**Comment by juanpicado** [comment]
No problem, I will take care of that 👍

File: assets/registry/store/@linclark/pkg/package.json (hunk 3 above)
Comment (juanpicado) [reply assets/registry/store/@linclark/pkg/package.json]: The files were just swapped, I fixed that.

**Comment by juanpicado** [comment]
Sorry the delay, I revisited all change requested, ready for next one :)

File: lib/registry.js [hunk 4]
```diff
@@ -41,265 +38,33 @@ function kill () {
...
-  var server = http.createServer(handler(args))
-  server.listen(15443, function () {
-    log('%s Listening', lesson)
-  })
-}
-
-function handler (args) {
+  var verdaccio = require('verdaccio').default
+  var yalm = require('js-yaml')
+  var ddd = path.join(__dirname, './_config.yaml')
```
Comment (ccarruitero) [inline lib/registry.js]: I think could be better use a more descriptive variable name here.
Maybe something like `yamlPath` or `yamlConfigPath`.

File: lib/registry.js [hunk 5]
```diff
@@ -41,265 +38,33 @@ function kill () {
...
-  var lesson = args[0]
-  var server = http.createServer(handler(args))
-  server.listen(15443, function () {
-    log('%s Listening', lesson)
-  })
-}
-
-function handler (args) {
+  var verdaccio = require('verdaccio').default
+  var yalm = require('js-yaml')
```
Comment (ccarruitero) [inline lib/registry.js]: We should move these module's require to top in file.

**Review by ccarruitero — CHANGES_REQUESTED** [review]
Thanks @juanpicado 

It's almost ready, just a minor changes.

File: package.json [hunk 6]
```diff
@@ -24,18 +24,18 @@
   },
   "bundleDependencies": [
     "workshopper-adventure",
-    "concat-stream",
```
Comment (juanpicado) [inline package.json:27]: @ccarruitero should I add the new dependencies here? 🤔

Comment (ccarruitero) [reply package.json:27]: I'm not sure that `bundleDependencies` are used for any process; but add the new dependency there, just in case.

**Review by ccarruitero — APPROVED** [review]

**Pull request merged**